odoo/addons/base/models/res_users.py

```diff
@@ -593,7 +593,21 @@ def write(self, values):
                 # safe fields only, so we write as super-user to bypass access rights
                 self = self.sudo().with_context(binary_field_real_user=self.env.user)
 
+        if 'groups_id' in values:
+            default_user = self.env.ref('base.default_user', raise_if_not_found=False)
+            if default_user and default_user in self:
+                old_group_ids = default_user.groups_id.ids
+
         res = super(Users, self).write(values)
+
+        if 'groups_id' in values and default_user and default_user in self:
+            # Sync added groups on default user template to existing users
+            new_group_ids = default_user.groups_id.ids
+            to_add_group_ids = list(set(new_group_ids) - set(old_group_ids))
+            if to_add_group_ids:
+                (self.env.ref('base.group_user').users - default_user).write({
+                    'groups_id': [(4, gid) for gid in to_add_group_ids]})
```

Comment:
Using `set` here introduces a source of nondeterminism, which has proven to be a source of painful bugs. And what's this number 4 for? :rofl: I strongly suggest to use recordsets instead, as they makes the code simpler and more reliable (no nondeterminism).
```suggestion
            added_groups = default_user.groups_id - old_groups
            if added_groups:
                internal_users = self.env.ref('base.group_user').users - default_user
                internal_users.write({'groups_id': [Command.link(gid) for gid in added_groups.ids]})
```
where `old_groups` is like `old_group_ids` without the `.ids`.